be/src/olap/rowset/segment_v2/binary_prefix_page.cpp

```diff
@@ -0,0 +1,276 @@
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "olap/rowset/segment_v2/binary_prefix_page.h"
+
+#include <map>
+#include <vector>
+#include <stddef.h>
+#include <stdint.h>
+#include <string.h>
+#include <string>
+
+#include "common/logging.h"
+#include "gutil/strings/substitute.h"
+#include "runtime/mem_pool.h"
+#include "util/coding.h"
+#include "util/faststring.h"
+#include "util/slice.h"
+
+namespace doris {
+namespace segment_v2 {
+
+using strings::Substitute;
+
+Status BinaryPrefixPageBuilder::add(const uint8_t* vals, size_t* add_count) {
+    DCHECK(!_finished);
+    if (*add_count == 0) {
+        return Status::OK();
+    }
+
+    const Slice* src = reinterpret_cast<const Slice*>(vals);
+    if (_count == 0) {
+        _first_entry.assign_copy(reinterpret_cast<const uint8_t*>(src->get_data()), src->get_size());
+    }
+
+    int i = 0;
+    for (; i < *add_count; ++i, ++src) {
+        if (is_page_full()) {
+            break;
+        }
+        const char* entry = src->data;
+        size_t entry_len = src->size;
+        int old_size = _buffer.size();
+
+        int share_len;
+        if (_count % RESTART_POINT_INTERVAL == 0) {
+            share_len = 0;
+            _restart_points_offset.push_back(old_size);
+        } else {
+            int max_share_len = std::min(_last_entry.size(), entry_len);
+            share_len = max_share_len;
+            for (int i = 0; i < max_share_len; ++i) {
+                if (entry[i] != _last_entry[i]) {
+                    share_len = i;
+                    break;
+                }
+            }
+        }
+        int non_share_len = entry_len - share_len;
+
+        put_varint32(&_buffer, share_len);
+        put_varint32(&_buffer, non_share_len);
+        _buffer.append(entry + share_len, non_share_len);
+
+        _last_entry.clear();
+        _last_entry.append(entry, entry_len);
+
+        ++_count;
+    }
+    *add_count = i;
+    return Status::OK();
+}
+
+OwnedSlice BinaryPrefixPageBuilder::finish() {
+    DCHECK(!_finished);
+    _finished = true;
+    put_fixed32_le(&_buffer, (uint32_t)_count);
+    uint8_t restart_point_internal = RESTART_POINT_INTERVAL;
+    _buffer.append(&restart_point_internal, 1);
+    auto restart_point_size = _restart_points_offset.size();
+    for(uint32_t i = 0; i < restart_point_size; ++i) {
+        put_fixed32_le(&_buffer, _restart_points_offset[i]);
+    }
+    put_fixed32_le(&_buffer, restart_point_size);
+    return _buffer.build();
+}
+
+const uint8_t* BinaryPrefixPageDecoder::_decode_value_lengths(const uint8_t* ptr,
+                                                              uint32_t* shared,
+                                                              uint32_t* non_shared) {
+    const uint8_t* limit = _restarts_ptr - sizeof(uint32_t);
```

Comment:
forget to minus 1(RESTART_POINT_INTERVAL).
Maybe we can use a member to record _limit to avoid compute limit every time.

Reply:
OK